store/postgres/src/store.rs

```diff
@@ -231,6 +234,76 @@ impl BasicStore for Store {
     }
 }
 
+impl BlockStore for Store {
+    fn add_network_if_missing(&self, network_name: &str) -> Result<(), Error> {
+        use db_schema::ethereum_networks::dsl::*;
+
+        insert_into(ethereum_networks)
+            .values((
+                name.eq(network_name),
+                head_block_hash.eq::<Option<String>>(None),
+                head_block_number.eq::<Option<i64>>(None),
+            ))
+            .on_conflict(name)
+            .do_nothing()
+            .execute(&self.conn)?;
+
+        Ok(())
+    }
+
+    fn upsert_blocks(&self, net_name: &str, blocks: &[Block<Transaction>]) -> Result<(), Error> {
+        use db_schema::ethereum_blocks::dsl::*;
+
+        for block in blocks {
+            let json_blob = serde_json::to_value(&block).expect("Failed to serialize block");
+            let values = (
+                hash.eq(format!("{:#x}", block.hash.unwrap())),
+                number.eq(block.number.unwrap().as_u64() as i64),
+                parent_hash.eq(format!("{:#x}", block.parent_hash)),
+                network_name.eq(net_name),
+                data.eq(json_blob),
+            );
+
+            insert_into(ethereum_blocks)
+                .values(values.clone())
+                .on_conflict(hash)
+                .do_update()
+                .set(values)
+                .execute(&self.conn)?;
+        }
+
+        Ok(())
+    }
+
+    fn attempt_head_update(
+        &self,
+        network_name: &str,
+        ancestor_count: u64,
+    ) -> Result<Vec<H256>, Error> {
+        select(attempt_head_update(network_name, ancestor_count as i64))
```

Comment:
Where does the `select` here come from?

Reply:
Added some comments on this. It's the way to run a stored SQL function in diesel (`attempt_head_update` is the stored function).

Reply:
Ah, cool, thanks.